text/0000-deployment-language.md

```diff
@@ -0,0 +1,561 @@
+#!config args my_arg = "foo", required_arg
+...
+```
+
+### Flow
+
+Flows are a new type of artefact that incorporates the previous concepts of `binding` and `mapping`.
+Flows define how artefacts are connected to form an event flow from sources via pipelines towards sinks.
+
+EBNF Grammar:
+
+```
+FLOW          = "define" "flow" flow_id
+                [ "args" argument_list ]
+                "links"
+                  link_list
+                "end"
+flow_id       = id
+argument_list = ( id | id "=" expr ) [ "," argument_list ]
+link_list     = link ";" [ link_list ]
+link          = "connect" tremor_url "to" tremor_url
+instance_port = artefact_id [ "/" instance_id ] ":" port
+```
+
+Example:
+
+```
+define flow my_eventflow
+args
+    required_arg,
+    optional_arg = "default value"
+links
+    connect "tremor://onramp/my_source/{required_arg}:out" to my_pipeline:in;
+    connect my_pipeline:out to my_sink:in;
+end;
+```
+
+#### Connect Statements
+
+Flows consists of `connect` statements, which connect two or more artefact instances via their ports. E.g. every instance
+that receives events has an `in` port. Event go out via the `out` port and error events are usually sent via the `err` port.
+A connect statement defines a connection between two instance-port references. These are provided as strings containing tremor urls
+in the following variants:
+
+* with instance-id and port (e.g. "/pipeline/my_pipeline/instance_id:in" )
+* without instance-id but with port (e.g. "/pipeline/my_pipeline:in" )
+* without instance id or port (e.g. "/pipeline/my_pipeline" )
+
+If no instance id is given, as part of the flow instantiation a new instance of the artefact is created with the flow instance id filled in as instance id. So, users can omit an instance id if they want to have their artefacts be auto-created as part of an event flow.
+They will also be auto-deleted upon flow deletion. Our goal with this setup is to make it possible for users to define their tremor setup
+as one unit, the `Flow`. It can be instantiated and deleted with one command or API call.
+
+If an already existing artefact instance is referenced via an instance-port reference in a connect statement, the `Flow` will not take ownership of those instances. It will just release the connections that are defined in the `links` section of its definition.
+With this logic, users are able to dynamically create and manage connections between existing artefacts separately from those artefacts itself.
+
+For this to work the `Flow` instance needs to track which instances it created in order to destroy them upon the process of itself being destroyed.
+
+If the port is ommitted, based on its position a default port is chosen. If a reference is on the LHS the port `out` is chosen, on the RHS `in` is implied. This matches the normal event flow: From the `out` port of an artefact to an `in` port of another one. Other cases need to be handled explicitly.
+
+#### Connect Statement Sugar
```

Comment:
Can we consider not having to quite the parts in the arrow sytnax? Something like:

```

# without ports
/source/stdin -> /pipeline/pipe;
/pipeline/pipe/err -> /sink/system::stderr;
# with ports
pipeline/pipe:err -> /sink/my_error_file:in;
```

Reply:
Yeah, i think we should discuss a non-stringy representation of TremorURL.